acra/acra-in-depth/data-structures/_index.md

```diff
@@ -2,3 +2,226 @@
+* `Rest_AcraBlock_Length[8]` - 8 bytes in Little-Endian byte order that store length of whole AcraBlock excluding length of Begin_Tag.
+* `Key_Encryption_Backend_Identifier[1]` - identifier of encryption backend used to encrypt DEK which is random symmetric key. For now AcraBlock supports only one backend that uses Themis Secure Cell in Seal mode with identifier `0` and may be changed.
+* `Key_Encryption_Key_ID[2]` - this field stores KEK's identifier. Main goal of this ID is to simplify search of KEK in a keychain related to specific ClientID/ZoneID. Now Acra-Server generates ID by hashing with `SHA256(KEK || ClientID/ZoneID)` and uses first 2 bytes as ID.
+* `Data_Encryption_Backend_Identifier[1]` -  identifier of encryption backend used to encrypt plaintext. For now AcraBlock supports only one backend that uses Themis Secure Cell in Seal mode with identifier `0`. Same as for DEK encryption and may be changed too.
+* `Data_Encryption_Key_Length[2]` - 2 bytes of DEK's length in Little-Endian byte order. Key size should be synchronized with data encryption backend.
+* `Encrypted_Data_Encryption_Key[*]` - encrypted DEK of size `Data_Encryption_Key_Length`, used to encrypt data and data encryption backend. For each new AcraBlock Acra-Server generates new random DEK and encrypts with DEK encryption backend and Context (ClientID/ZoneID).
+* `Encrypted_Data[*]` encrypted data with data encryption backend and random DEK.
+
+Now exists only one key and data decryption backends that use Secure Cell in Seal Mode.
+
+Available **key** encryption backends and their identifiers:
+* `Secure Cell in Seal Mode` - `0`
+
+Available **data** encryption backends and their identifiers:
+* `Secure Cell in Seal Mode` - `0`
+
+
+AcraBlock example for plaintext: `example`:
+
+`[34 34 34 34 141 0 0 0 0 0 0 0 0 119 199 0 76 0 0 1 1 64 12 0 0 0 16 0 0 0 32 0 0 0 74 57 33 205 212 208 201 173 5 49 141 175 138 74 129 21 72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 3 92 233 23 106 136 192 110 24 237 45 62 109 105 70]`
+
+* Begin_Tag[4] - `[34 34 34 34]`
+* Rest_AcraBlock_Length[8] - `[141 0 0 0 0 0 0 0]`
+* Key_Encryption_Backend_Identifier[1] - `[0]`
+* Key_Encryption_Key_ID[2] - `[119 199]`
+* Data_Encryption_Backend_Identifier[1] - `[0]`
+* Data_Encryption_Key_Length[2] - `[76 0]`
+* Encrypted_Data_Encryption_Key[*] - `[0 1 1 64 12 0 0 0 16 0 0 0 32 0 0 0 74 57 33 205 212 208 201 173 5 49 141 175 138 74 129 21]`
+* Encrypted_Data[*] - `[72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 3 92 233 23 106 136 192 110 24 237 45 62 109 105 70]`
```

Comment:
IMO: For me it looks normally (don't know about typical user). Often it is required to analyze the output of raw bytes, and I'm accustomed seeing that say DataLength is 8 bytes - [51, 0, 0, 0, 0, 0, 0, 0] and they are definitely can by found in overall AcraStruct stream of bytes

Reply:
It just looks sloppy. Like Acra developer just copied and pasted debugger output here, without giving any thought of how it would look, how it would read, how non-Acra-developers are going to make any sense of it, and what is the overall goal of this documentation in the first place. This checks the box “we have N kilobytes of text that explains something about AcraStruct internals” but not every text is documentation, let alone *good* documentation.

`[72 124 140 56 23 187 147 90 79 15 192 32 164 252 144 234 120 9 212 13 105 61 48 221 98 14 188 170 171 97 212 137 231 11 99 218 252 110 120 212 32 72 195 54 0 1 1 64 12 0 0 0 16 0 0 0 7 0 0 0 72 81 94 231 179 35 95 35 89 97 196 52 15 142 126 144 136 26 86 192 3 92 233 23 106 136 192 110 24 237 45 62 109 105 70]` is no more informative than “95 bytes”.